### PR TITLE
Base: Add handling of additional units used in CfdOF workbench

### PR DESCRIPTION
**Pull request by oliveroxtoby**
This adds handling for some unusual units which are used in the CfdOF workbench, to prevent them being displayed in unconventional units or with strange scalings, e.g. mm^2/s^3 for the turbulent dissipation rate. This is just a straightforward extension of what is done for other units in the schemas.

Forum discussion: https://forum.freecadweb.org/viewtopic.php?f=37&t=21576&p=585747&sid=2117aa796c409dd139b76e26ad8a0712#p584248 and posts following.

- [x]  Your pull request is confined strictly to a single module. That is, all the files changed by your pull request are either in `App`, `Base`, `Gui` or one of the `Mod` subfolders. If you need to make changes in several locations, make several pull requests and wait for the first one to be merged before submitting the next ones
- [x]  In case your pull request does more than just fixing small bugs, make sure you discussed your ideas with other developers on the FreeCAD forum
- [x]  Your branch is [rebased](https://git-scm.com/docs/git-rebase) on latest master `git pull --rebase upstream master`
- [x] (some unrelated failures)  All FreeCAD unit tests are confirmed to pass by running `./bin/FreeCAD --run-test 0`
- [x]  All commit messages are [well-written](https://chris.beams.io/posts/git-commit/) ex: `Fixes typo in Draft Move command text`
- [x]  Your pull request is well written and has a good description, and its title starts with the module name, ex: `Draft: Fixed typos`
- [ ]  Commit messages include `issue #<id>` or `fixes #<id>` where `<id>` is the issue ID number from our [Issues database](https://github.com/FreeCAD/FreeCAD/issues) in case a particular commit solves or is related to an existing issue. Ex: `Draft: fix typos - fixes #4805`

And please remember to update the Wiki with the features added or changed once this PR is merged.  
**Note**: If you don't have wiki access, then please mention your contribution on the [0.20 Changelog Forum Thread](https://forum.freecadweb.org/viewtopic.php?f=10&t=56135).

---

### CODE REVIEW

**Comment by freecadci** [comment]
<a href="https://gitlab.com/freecad/FreeCAD-CI/-/commits/PR_6717">pipeline status</a> for feature branch [PR_6717](https://gitlab.com/freecad/FreeCAD-CI/-/commits/PR_6717). Pipeline [510470558](https://gitlab.com/freecad/FreeCAD-CI/-/pipelines/510470558) was triggered at [aa7a5fb](https://github.com/FreeCAD/FreeCAD/pull/6717/commits/aa7a5fb01e536ced18c219e5b00831e872091115). All CI [branches](https://gitlab.com/freecad/FreeCAD-CI/-/branches/all) and [pipelines](https://gitlab.com/freecad/FreeCAD-CI/-/pipelines?scope=branches).

**Comment by donovaly** [comment]
Thanks for the PR. However, this is a feature addition and your PR covers only part of the necessary changes.

See as reference this commit: https://github.com/FreeCAD/FreeCAD/commit/bcc669b930

However, for the workbench, such a PR is even not complete since you need to make a new App::Property and use it for the parameters. For example like it is done in this commit:  https://github.com/FreeCAD/FreeCAD/commit/d286b52b57

**Comment by donovaly** [comment]
Since it is on my ToDo list, to improve the unit handling for FEM in general, I will do something after FreeCAD 0.20 was released.

**Comment by oliveroxtoby** [comment]
> Thanks for the PR. However, this is a feature addition and your PR covers only part of the necessary changes.
> 
> See as reference this commit: [bcc669b930](https://github.com/FreeCAD/FreeCAD/commit/bcc669b930)

Thanks for your time and I am sorry that this doesn't meet your expectations. I don't want to be difficult, but I am struggling to understand what I have done wrong here.

For the first commit that you refer to ([bcc669b930](https://github.com/FreeCAD/FreeCAD/commit/bcc669b930)), if I go through the changes I see the following:
- The name of a new quantity (ElectricalConductivity in your case) added to Units.h and its dimensions defined in Units.cpp. This is the same as I have done for the new quantites I added - VolumeFlowRate, DissipationRate, InverseLength, InverseArea, InverseVolume.
- Handling of the scaling of these new quantites added to UnitsSchemaInternal.cpp and UnitsSchemaMKS.cpp, just as I have done for the quantities mentioned above.
- Names of new _units_ added to Quantity (KiloSiemens and MegaSiemens) and their new symbols defined (kS, MS) in the lexer (ExpressionParser.l) file. However the new quantites I have added do not require any new named units at all. They are purely expressed in terms of existing units (m^2/s^3, 1/m, 1/m^2, 1/m^3). I cannot see what there is to add here. Perhaps I am missing something.
- The new quantities added to the units calculator tool (DlgUnitsCalculatorImp.cpp). Should I have added the new quantities here? The reasons I did not do so are:
    - If these quantites are only used in an add-on workbench, is it right that they are user-facing in the core code? I just assumed I should not pollute it with esoteric quantities which only appear in the add-on.
    - Of the 44 quantites currently declared in Unit.h, the following are not present in the units calculator: CompressiveStrength, DynamicViscosity, KinematicViscosity, ShearModulus, SpecificEnergy, UltimateTensileStrength, VacuumPermittivity, VolumetricThermalExpansionCoefficient, YieldStrength, YoungsModulus. This led me to believe that the more 'esoteric' units were not included here. Do you want me to add all the missing ones?
- Material properties (ElectricalConductivity and RelativePermittivity) are added to the material properties template.  I do not believe this is relevant here as the none of the added quantities are material properties.

> 
> However, for the workbench, such a PR is even not complete since you need to make a new App::Property and use it for the parameters. For example like it is done in these 2 commits: [d286b52b57](https://github.com/FreeCAD/FreeCAD/commit/d286b52b57) and [30928c071](https://github.com/FreeCAD/FreeCAD/commit/30928c071)

I'm very happy to add this if you want it, but this does not seem clear-cut to me at all. Of the 44 quantities mentioned before, I can only find 10 of them for which an App::Property... has been defined in Application.cpp. This led me to belive that only certain 'core' quantities were given their own dedicated property. I don't see why one can't just use a App::PropertyQuantity for these, which is what I do in CfdOF for all the quantites that don't have their own property already defined. It works perfectly fine, and prevents me having to make requests for core changes to accommodate all kinds of wierd quantities which are only used in an addon. However I may be missing some advantage of this approach and am always happy to be corrected.

To be clear, I have tested and found no problems with the use of these quantities as I have them in this pull request. The workbench is working as expected and having no problems displaying and parsing them in the InputFields. For example:

Before:
Screenshot from 2022-04-13 08-16-26
After:
Screenshot from 2022-04-13 08-17-42

Before:
Screenshot from 2022-04-13 08-21-59
After:
Screenshot from 2022-04-13 08-18-26

Thank you for your patience.

**Comment by donovaly** [comment]
Your PR not bad, so it is not that it did not fit to the expectations. I only think it is incomplete.

Of course not everything of the example commits I referred to applies for your case. For combined units like "ml/min", no lexer entry is necessary.

That some units are missing in the DlgUnitsCalculatorImp is a matter of lazyness I guess. I think all units should be supported there. Can you please modify the PR accordingly?

> I can only find 10 of them for which an App::Property... has been defined in Application.cpp

App::Property... is only necessary to use the unit for either an object property or a task dialog field. Since you want to use the new units in a task dialog for the WB, an  App::Property... is necessary.
Note that this only applies for cases where you want a certain output. For example you want the output "ml/s" and not mm^3/s. 
Maybe I am wrong since I did not yet test if ml is otherwise translated to mm³.
Can you please check this?

We are in feature freeze and the PR cannot go in right now.

**Comment by freecadci** [comment]
<a href="https://gitlab.com/freecad/FreeCAD-CI/-/commits/PR_6717">pipeline status</a> for feature branch [PR_6717](https://gitlab.com/freecad/FreeCAD-CI/-/commits/PR_6717). Pipeline [533998541](https://gitlab.com/freecad/FreeCAD-CI/-/pipelines/533998541) was triggered at [0f77551](https://github.com/FreeCAD/FreeCAD/pull/6717/commits/0f775519cdb84bb5be0545b7f90b4ad95232d61c). All CI [branches](https://gitlab.com/freecad/FreeCAD-CI/-/branches/all) and [pipelines](https://gitlab.com/freecad/FreeCAD-CI/-/pipelines?scope=branches).

**Comment by oliveroxtoby** [comment]
> That some units are missing in the DlgUnitsCalculatorImp is a matter of lazyness I guess. I think all units should be supported there. Can you please modify the PR accordingly?
> 

OK, I have done that for all the units except for:
CompressiveStrength
ShearModulus
Stress
UltimateTensileStrength
YieldStrength
YoungsModulus
These all show up in the units calculator as 'Pressure' since they share the same units as Pressure, so I have removed them. (They can't be distinguisehd since Unit objects are presently identified by their fundamental dimensions only, so these objects are all identical to each other.)

> App::Property... is only necessary to use the unit for either an object property or a task dialog field. Since you want to use the new units in a task dialog for the WB, an App::Property... is necessary. Note that this only applies for cases where you want a certain output. For example you want the output "ml/s" and not mm^3/s. Maybe I am wrong since I did not yet test if ml is otherwise translated to mm³. Can you please check this?

As I mentioned, it is working correctly for me in the task dialog when simply using PropertyQuantity (for example, the volume flow rate displaying as 'ml/s', not mm^3/s). I tested as you ask and I see no difference in the output if I define a special property (e.g. PropertyVolumeFlowRate in this example). This makes sense to me because the InputField uses Quantity for format the output,  and this is purely based on the Units assigned to the quantity and the representation of those units defined in the UnitsSchema, and doesn't involve the Property object.

I nonetheless have updated the PR to add Property objects for all the outstanding units, for completeness. Please let me know if this is not what you want.

> We are in feature freeze and the PR cannot go in right now.

No problem for me, I was not in a rush with this. I just wanted to put in the request.

Finally, I have added one other commit that I thought fitted with this discussion: I changed the units displayed for dynamic viscosity. Presently this is displayed as kg/(mm*s) in Internal and kg/(m*s) in MKS. However it is more common to use the equivalent units of Pa*s, and I have proposed that it be changed to this in both cases. kg/(mm*s) is certainly not a scaling that I have ever seen used anywhere.

**Comment by freecadci** [comment]
<a href="https://gitlab.com/freecad/FreeCAD-CI/-/commits/PR_6717">pipeline status</a> for feature branch [PR_6717](https://gitlab.com/freecad/FreeCAD-CI/-/commits/PR_6717). Pipeline [534009738](https://gitlab.com/freecad/FreeCAD-CI/-/pipelines/534009738) was triggered at [7e4c5c1](https://github.com/FreeCAD/FreeCAD/pull/6717/commits/7e4c5c1f7216a5954f87ddd61b25132b46bb8b6e). All CI [branches](https://gitlab.com/freecad/FreeCAD-CI/-/branches/all) and [pipelines](https://gitlab.com/freecad/FreeCAD-CI/-/pipelines?scope=branches).

**Comment by luzpaz** [comment]
Can work on this ticket restart ?

**Comment by donovaly** [comment]
> Can work on this ticket restart ?

Yes.

**Comment by luzpaz** [comment]
Ping @oliveroxtoby

**Comment by freecadci** [comment]
<a href="https://gitlab.com/freecad/FreeCAD-CI/-/commits/PR_6717">pipeline status</a> for feature branch [PR_6717](https://gitlab.com/freecad/FreeCAD-CI/-/commits/PR_6717). Pipeline [700153850](https://gitlab.com/freecad/FreeCAD-CI/-/pipelines/700153850) was triggered at [a3ac4e8](https://github.com/FreeCAD/FreeCAD/pull/6717/commits/a3ac4e8f92c3f6b8b814ffbaf08eb773c0389302). All CI [branches](https://gitlab.com/freecad/FreeCAD-CI/-/branches/all) and [pipelines](https://gitlab.com/freecad/FreeCAD-CI/-/pipelines?scope=branches).

**Comment by oliveroxtoby** [comment]
> Ping @oliveroxtoby

I have now rebased on the latest master, if this is what you were getting at?

**Comment by freecadci** [comment]
<a href="https://gitlab.com/freecad/FreeCAD-CI/-/commits/PR_6717">pipeline status</a> for feature branch [PR_6717](https://gitlab.com/freecad/FreeCAD-CI/-/commits/PR_6717). Pipeline [700179684](https://gitlab.com/freecad/FreeCAD-CI/-/pipelines/700179684) was triggered at [d23de8e](https://github.com/FreeCAD/FreeCAD/pull/6717/commits/d23de8e295bdc18240f354b3f9693c5e83be654f). All CI [branches](https://gitlab.com/freecad/FreeCAD-CI/-/branches/all) and [pipelines](https://gitlab.com/freecad/FreeCAD-CI/-/pipelines?scope=branches).

**Comment by freecadci** [comment]
<a href="https://gitlab.com/freecad/FreeCAD-CI/-/commits/PR_6717">pipeline status</a> for feature branch [PR_6717](https://gitlab.com/freecad/FreeCAD-CI/-/commits/PR_6717). Pipeline [700193332](https://gitlab.com/freecad/FreeCAD-CI/-/pipelines/700193332) was triggered at [b9bae2f](https://github.com/FreeCAD/FreeCAD/pull/6717/commits/b9bae2fc84cd5790f8c92e009742dd628dd99bcc). All CI [branches](https://gitlab.com/freecad/FreeCAD-CI/-/branches/all) and [pipelines](https://gitlab.com/freecad/FreeCAD-CI/-/pipelines?scope=branches).

**Comment by luzpaz** [comment]
@oliveroxtoby any traction on this ?

**Comment by chennes** [comment]
The PR is still marked as a draft -- once the conflicts have been resolved, if it's ready to be merged, make sure to switch it to  "Ready for Review".

**Comment by freecadci** [comment]
<a href="https://gitlab.com/freecad/FreeCAD-CI/-/commits/PR_6717">pipeline status</a> for feature branch [PR_6717](https://gitlab.com/freecad/FreeCAD-CI/-/commits/PR_6717). Pipeline [782414455](https://gitlab.com/freecad/FreeCAD-CI/-/pipelines/782414455) was triggered at [727d58e](https://github.com/FreeCAD/FreeCAD/pull/6717/commits/727d58ea46a85334a38ee4fb77b29b693814e6c3). All CI [branches](https://gitlab.com/freecad/FreeCAD-CI/-/branches/all) and [pipelines](https://gitlab.com/freecad/FreeCAD-CI/-/pipelines?scope=branches).

**Comment by oliveroxtoby** [comment]
> @oliveroxtoby any traction on this ?

@luzpaz I'm afraid you'll have to ask @donovaly, who had some issues with this pull request initially and marked it as Draft. I hope you can see from the above that I have done my best to accommodate the requests, but I really don't know where we stand with this now. In any case, I have rebased once again and it is still good to go from my point of view.

**Comment by donovaly** [comment]
> @luzpaz I'm afraid you'll have to ask @donovaly, who had some issues with this pull request initially and marked it as Draft.

We were that time in feature freeze and the PR was incomplete. If you think now everything is in shape, please undraft it and I or other developers will review it.

File: src/Base/UnitsSchemaInternal.cpp [hunk 1]
```diff
@@ -480,7 +480,7 @@ QString UnitsSchemaInternal::schemaTranslate(const Quantity &quant, double &fact
         factor = 1.0;
     }
     else if (unit == Unit::DynamicViscosity) {
-        unitString = QString::fromLatin1("kg/(mm*s)");
+        unitString = QString::fromLatin1("Pa*s");
```
Comment (donovaly) [inline src/Base/UnitsSchemaInternal.cpp:483]: 1 kg/(mm*s ) = 1000 Pa*s With this change it can be you change the values for existing documents.
However, I support using SI units but this change needs to be checked carefully with existing documents.

File: src/Base/UnitsSchemaMKS.cpp [hunk 2]
```diff
@@ -470,7 +470,7 @@ QString UnitsSchemaMKS::schemaTranslate(const Quantity &quant, double &factor, Q
         factor = 1000.0;
     }
     else if (unit == Unit::DynamicViscosity) {
-        unitString = QString::fromLatin1("kg/(m*s)");
+        unitString = QString::fromLatin1("Pa*s");
```
Comment (donovaly) [inline src/Base/UnitsSchemaMKS.cpp:473]: same here

File: src/App/PropertyUnits.cpp [hunk 3]
```diff
@@ -357,6 +357,79 @@ PropertyStiffness::PropertyStiffness()
...
+TYPESYSTEM_SOURCE(App::PropertyMass, App::PropertyQuantity)
+
+PropertyMass::PropertyMass()
+{
+    setUnit(Base::Unit::Mass);
+}
+
+//**************************************************************************
+//**************************************************************************
+// PropertyAngleOfFriction
```
Comment (donovaly) [inline src/App/PropertyUnits.cpp:374]: Why is the existing PropertyAngle not usable? I think we should use the existing properties.

File: src/App/PropertyUnits.cpp [hunk 4]
```diff
@@ -357,6 +357,79 @@ PropertyStiffness::PropertyStiffness()
...
+TYPESYSTEM_SOURCE(App::PropertyDensity, App::PropertyQuantity)
+
+PropertyDensity::PropertyDensity()
+{
+    setUnit(Base::Unit::Density);
+}
+
+//**************************************************************************
+//**************************************************************************
+// PropertyTimeSpan
```
Comment (donovaly) [inline src/App/PropertyUnits.cpp:399]: Here I would use a generic property "Time". This can then be used for all kinds of times with the SI unit s.

File: src/Gui/DlgUnitsCalculatorImp.cpp [hunk 5]
```diff
@@ -86,27 +88,34 @@ DlgUnitsCalculator::DlgUnitsCalculator( QWidget* parent, Qt::WindowFlags fl )
...
+          << Base::Unit::InverseVolume
+          << Base::Unit::KinematicViscosity
           << Base::Unit::Length
           << Base::Unit::LuminousIntensity
           << Base::Unit::Mass
           << Base::Unit::MagneticFieldStrength
           << Base::Unit::MagneticFlux
           << Base::Unit::MagneticFluxDensity
           << Base::Unit::Magnetization
-          << Base::Unit::Pressure
```
Comment (donovaly) [inline src/Gui/DlgUnitsCalculatorImp.cpp:100]: Why did you delete this?

File: src/Gui/DlgUnitsCalculatorImp.cpp [hunk 6]
```diff
@@ -86,27 +88,34 @@ DlgUnitsCalculator::DlgUnitsCalculator( QWidget* parent, Qt::WindowFlags fl )
...
           << Base::Unit::Power
+          << Base::Unit::Pressure
+          << Base::Unit::SpecificEnergy
           << Base::Unit::SpecificHeat
-          << Base::Unit::Stress
```
Comment (donovaly) [inline src/Gui/DlgUnitsCalculatorImp.cpp:103]: Why did you delete this?

File: src/Base/Unit.cpp [hunk 7]
```diff
@@ -572,3 +582,9 @@ Unit Unit::HeatFlux                    (0,1,-3,0,0);
 Unit Unit::DynamicViscosity            (-1,1,-1);  // SI unit: kg/m/s
 Unit Unit::KinematicViscosity          (2,0,-1);  // SI unit: m^2/s, https://en.wikipedia.org/wiki/Viscosity#Kinematic_viscosity
 Unit Unit::VacuumPermittivity          (-3,-1,4,2); // SI unit: A²*s⁴/kg/m³ https://en.wikipedia.org/wiki/Permittivity#Vacuum_permittivity
+Unit Unit::VolumeFlowRate              (3,0,-1);
+Unit Unit::DissipationRate             (2,0,-3);  // SI unit: m^2/s^3 https://cfd-online.com/Wiki/Turbulence_dissipation_rate
```
Comment (donovaly) [inline src/Base/Unit.cpp:598]: I think the best is to just use  the Wiki link

File: src/Base/Unit.cpp [hunk 8]
```diff
@@ -508,6 +508,16 @@ QString Unit::getTypeString() const
         return QString::fromLatin1("KinematicViscosity");
     if (*this == Unit::VacuumPermittivity)
         return QString::fromLatin1("VacuumPermittivity");
+    if(*this == Unit::VolumeFlowRate              )
```
Comment (donovaly) [inline src/Base/Unit.cpp:523]: Minor but nonetheless: please follow the existing formatting of a file. This might seem nitpicking but later the automated clang formatter will produce unnecessary diffs when someone change s the code in future.

**Review by donovaly — CHANGES_REQUESTED** [review]
Man thanks for the PR! Very valuable addition also for example for the Elmer solver.

For the angle and time, we should therefore use a generic property to be used later by other devs for all kinds of angles and times.

File: src/Base/UnitsSchemaInternal.cpp (hunk 1 above)
Comment (oliveroxtoby) [reply src/Base/UnitsSchemaInternal.cpp:483]: Unless I'm mistaken the quantities are always stored in FreeCAD internal units on file, so there should be no problem with conversion. To be sure I have saved a document with the old code opened with the new and indeed it is correctly represented in the new version.

File: src/App/PropertyUnits.cpp (hunk 3 above)
Comment (oliveroxtoby) [reply src/App/PropertyUnits.cpp]: Sure, PropertyAngle is usable, but I don't know how to associate both Angle and AngleOfFriction units with a single property (and I don't see that it would make sense). Probably we should just not define a property for AngleOfFriction, so I have removed it.

File: src/App/PropertyUnits.cpp (hunk 4 above)
Comment (oliveroxtoby) [reply src/App/PropertyUnits.cpp]: OK, changed.

File: src/Gui/DlgUnitsCalculatorImp.cpp (hunk 5 above)
Comment (oliveroxtoby) [reply src/Gui/DlgUnitsCalculatorImp.cpp:100]: I didn't; I moved it two lines down. Almost all the other quantities are in alphapetical order here and this one was slightly out of position.

File: src/Gui/DlgUnitsCalculatorImp.cpp (hunk 6 above)
Comment (oliveroxtoby) [reply src/Gui/DlgUnitsCalculatorImp.cpp:103]: It was for the reason I mentioned above:

> OK, I have done that for all the units except for:
> CompressiveStrength
> ShearModulus
> Stress
> UltimateTensileStrength
> YieldStrength
> YoungsModulus
> These all show up in the units calculator as 'Pressure' since they share the same units as Pressure, so I have removed them. (They can't be distinguished since Unit objects are presently identified by their fundamental dimensions only, so these objects are all identical to each other.)
> 
>

File: src/Base/Unit.cpp (hunk 7 above)
Comment (oliveroxtoby) [reply src/Base/Unit.cpp]: I couldn't find one, so I have removed the cfd-online link.

File: src/Base/Unit.cpp (hunk 8 above)
Comment (oliveroxtoby) [reply src/Base/Unit.cpp]: No, I agree; this is exactly the reason I used this funny formatting in the first place. However the formatting of the file was changed in the interim, and I missed this when rebasing; apologies.

**Comment by oliveroxtoby** [comment]
> We were that time in feature freeze and the PR was incomplete. If you think now everything is in shape, please undraft it and I or other developers will review it.

OK, I didn't realise that was the procedure. Felt it would be a bit presumptuous to undraft it myself.

**Comment by oliveroxtoby** [comment]
> Man thanks for the PR! Very valuable addition also for example for the Elmer solver.
> 
> For the angle and time, we should therefore use a generic property to be used later by other devs for all kinds of angles and times.

Thanks. Agreed.

**Comment by freecadci** [comment]
<a href="https://gitlab.com/freecad/FreeCAD-CI/-/commits/PR_6717">pipeline status</a> for feature branch [PR_6717](https://gitlab.com/freecad/FreeCAD-CI/-/commits/PR_6717). Pipeline [788883091](https://gitlab.com/freecad/FreeCAD-CI/-/pipelines/788883091) was triggered at [247467c](https://github.com/FreeCAD/FreeCAD/pull/6717/commits/247467c8f2789e51b80b9c303c309e1577daedd4). All CI [branches](https://gitlab.com/freecad/FreeCAD-CI/-/branches/all) and [pipelines](https://gitlab.com/freecad/FreeCAD-CI/-/pipelines?scope=branches).

File: src/App/PropertyUnits.cpp (hunk 3 above)
Comment (donovaly) [reply src/App/PropertyUnits.cpp]: Then you misunderstood me. What you did is per se not wrong. AngleOfFriction exists already but we should have a property just called "angle" that can be used by all WBs and all kinds of angles.

Let's do it so: keep your PR as it is, I will merge and add a generic angle in a further PR. OK?

File: src/Base/UnitsSchemaInternal.cpp (hunk 1 above)
Comment (donovaly) [reply src/Base/UnitsSchemaInternal.cpp:483]: Thanks for testing. So fine with me to make this change.

File: src/Base/Unit.cpp (hunk 7 above)
Comment (donovaly) [reply src/Base/Unit.cpp]: Oh, my mistake you already correctly use the Wiki link I mean. So keeping
https://www.cfd-online.com/Wiki/Turbulence_dissipation_rate
is correct. Please keep it.

File: src/Base/UnitsSchemaInternal.cpp (hunk 1 above)
Comment (oliveroxtoby) [reply src/Base/UnitsSchemaInternal.cpp:483]: Thanks

File: src/App/PropertyUnits.cpp (hunk 3 above)
Comment (oliveroxtoby) [reply src/App/PropertyUnits.cpp]: OK, I will leave as is. There is a property already called Angle, I thought that is what you meant.

File: src/Base/Unit.cpp (hunk 7 above)
Comment (oliveroxtoby) [reply src/Base/Unit.cpp:598]: OK, I have added it back.

File: src/Base/UnitsSchemaInternal.cpp (hunk 1 above)
Comment (oliveroxtoby) [reply src/Base/UnitsSchemaInternal.cpp:483]: Thanks

**Comment by freecadci** [comment]
<a href="https://gitlab.com/freecad/FreeCAD-CI/-/commits/PR_6717">pipeline status</a> for feature branch [PR_6717](https://gitlab.com/freecad/FreeCAD-CI/-/commits/PR_6717). Pipeline [789242949](https://gitlab.com/freecad/FreeCAD-CI/-/pipelines/789242949) was triggered at [44738b7](https://github.com/FreeCAD/FreeCAD/pull/6717/commits/44738b7431e32db45a20dc45bb80dfe3badcd305). All CI [branches](https://gitlab.com/freecad/FreeCAD-CI/-/branches/all) and [pipelines](https://gitlab.com/freecad/FreeCAD-CI/-/pipelines?scope=branches).

**Review by donovaly — APPROVED** [review]

**Comment by donovaly** [comment]
Thanks for your patience and your work. Let's put it in and fine-tune if necessary.
It is good that now all properties are also registered to be used in the GUI.

I will have a look at the generic Angle property the next days.

**Comment by wwmayer** [comment]
None of the new properties are registered. So, it won't be possible to use them in feature types.
This must be done in App::Application::initTypes()

**Comment by donovaly** [comment]
> None of the new properties are registered. So, it won't be possible to use them in feature types. This must be done in App::Application::initTypes()

I am already working on this. At first In had to fix the compilation: #8669

**Comment by oliveroxtoby** [comment]
> 

Thank you.

**Comment by oliveroxtoby** [comment]
> None of the new properties are registered. So, it won't be possible to use them in feature types. This must be done in App::Application::initTypes()

Sorry that I missed that.

**Comment by wwmayer** [comment]
In I have added some more comments: https://github.com/FreeCAD/FreeCAD/commit/9d5ab79a4c8bd92c9e4f0c7ed395abbc22c2683d

There are four places where a wrong unit is used. @donovaly if you are working in this one then please fix the other issues, too.

**Comment by donovaly** [comment]
> In I have added some more comments: [9d5ab79](https://github.com/FreeCAD/FreeCAD/commit/9d5ab79a4c8bd92c9e4f0c7ed395abbc22c2683d)
> 
> There are four places where a wrong unit is used. @donovaly if you are working in this one then please fix the other issues, too.

Done: #8678

**Comment by oliveroxtoby** [comment]
> In I have added some more comments: [9d5ab79](https://github.com/FreeCAD/FreeCAD/commit/9d5ab79a4c8bd92c9e4f0c7ed395abbc22c2683d)
> 
> There are four places where a wrong unit is used. @donovaly if you are working in this one then please fix the other issues, too.

That's embarrasing! Thanks for picking them up. And thanks @donovaly for sorting out.
